### DISCUSSION TITLE
tilelayer data from private GitHub repo using authorization token

### DISCUSSION THREAD

**Discussion by cornhundred** [visgl/deck.gl#5499] [Q&A] [answered]
Hi, would it be possible to have TileLayer's getTileData retrieve data from a private GitHub repo by passing an Authorization token? I got the moon image example working on ObservableHQ ([link](https://observablehq.com/@cornhundred/deck-gl-moon-tillayer-example)), but I would like to read the JPEG pyramid images from a private repo using something like this (below I'm loading a JSON from a private repo by passing the token using ObservableHQ's secrets).

```
private_data = axios({
      method: 'get',
      url: 'https://api.github.com/repos/cornhundred/private-repo/contents/example.json',
      headers: {
        'Authorization': 'token ' + Secret("secret-token"),
        'Accept': 'application/vnd.github.v3.raw',
      }
    })
    .then(function (response) {
      return response.data
    });
```

Would I be able to pass in the token using the existing loader or would I need to make a custom loader?

**Comment by Pessimistress** [comment]
[fetch data with credentials](https://deck.gl/docs/developer-guide/loading-data#example-fetch-data-with-credentials)

If you need per-tile handling or post processing of the data, you can also provide your own [getTileData](https://deck.gl/docs/api-reference/geo-layers/tile-layer#gettiledata) prop.

**Comment by kylebarron** [comment]
You should be able to pass `fetch` options to the [`loadOptions` prop](https://deck.gl/docs/api-reference/core/layer#loadoptions), like
```js
loadOptions: {fetch: {headers}}
```

**Comment by cornhundred** [comment]
Thanks for the quick help @Pessimistress and @kylebarron. I'll try that out and update here.

**Comment by cornhundred** [comment, marked as answer]
I tried adding `loadOptions` to the TileLayet, but I am getting a 404 error on the GET request (from the console). Am I setting the loadOptions correctly?

```
  const moonLayer = new deck.TileLayer({
      pickable: autoHighlight,
      tileSize: dimensions.tileSize,
      autoHighlight,
      highlightColor: [60, 60, 60, 100],
      minZoom: -7,
      maxZoom: 0,
      coordinateSystem: deck.COORDINATE_SYSTEM.CARTESIAN,
      extent: [0, 0, dimensions.width, dimensions.height],,
      getTileData: ({x, y, z}) => {
        return loaders.load(`${ROOT_URL}/moon.image_files/${15 + z}/${x}_${y}.jpeg`);
      },
      onViewportLoad: getMetaData,
      renderSubLayers: renderSubLayers,
      loadOptions: {
        fetch: {
          method: 'GET',
          headers: {
            'Authorization': 'token ' + Secret("secret-token"),
            'Accept': 'application/vnd.github.v3.raw',
          }
        }
      }
    
    });   

```

**Reply by ibgreen** [reply]
Looks right at first blush - check your networks tab to see if it propagates.

**Reply by cornhundred** [reply]
Thanks @ibgreen, the network tab shows the following

<img width="828" alt="Screen Shot 2021-02-18 at 9 53 05 PM" src="https://user-images.githubusercontent.com/8352840/108450839-d0fbe600-7233-11eb-8e76-878e5b7db662.png">

The original public example used the following URL 
```
ROOT_URL = 'https://raw.githubusercontent.com/cornhundred/deck.gl-data/master/website/image-tiles/moon.image'
```

while the private repo (the repo is private, but just for testing purposes) one uses the `api.github.com` URL (which as far as I understand is required for authorization)
```
ROOT_URL = 'https://api.github.com/repos/cornhundred/private-moon/contents/moon.image'
```

**Reply by ibgreen** [reply]
You can expand the request in the network tab and see which headers were actually sent.

<img width="537" alt="Screen Shot 2021-02-18 at 7 07 21 PM" src="https://user-images.githubusercontent.com/7025232/108452052-abafad80-721c-11eb-9ca3-0cb475f65104.png">

If they are not present, something is going wrong in deck. If they are present, then the issue is outside of deck, if so I recommend you use tools like `curl` or `postman` to experiment with the URL and headers until you get the API request working.

**Reply by kylebarron** [reply]
Especially since the error is 404, it seems likelier that the URL is incorrect. If the authentication headers weren't being sent to a valid URL, I'd expect a 401 Unauthorized or 403 Forbidden.

**Reply by cornhundred** [reply]
Thanks @ibgreen and @kylebarron, I see headers (see screenshot) so it sounds like the issue is outside deck. I'll try curl and postman 
<img width="1353" alt="Screen Shot 2021-02-19 at 10 13 27 AM" src="https://user-images.githubusercontent.com/8352840/108522862-3df69680-729b-11eb-9915-06a3833fd002.png">

**Reply by kylebarron** [reply]
> I see headers (see screenshot)

That screenshot cuts off before the request headers. The relevant part here is whether the _request_ headers are set correctly. If you set in code:
```js
loadOptions: {
  fetch: {
    method: 'GET',
    headers: {
      'Authorization': 'token ' + Secret("secret-token"),
      'Accept': 'application/vnd.github.v3.raw',
    }
  }
}
```
Then you should check whether the `Authorization` and `Accept` _request_ headers are included, not the response headers back from the server.

**Reply by cornhundred** [reply]
Sorry, here's the cutoff headers part

<img width="1096" alt="Screen Shot 2021-02-19 at 11 55 22 AM" src="https://user-images.githubusercontent.com/8352840/108535583-7b622080-72a9-11eb-8f4b-c691b55e5157.png">

Thanks, for the help @kylebarron - I'll try to look into what you're describing

**Reply by ibgreen** [reply]
I don't see the authorization/accept headers coming through. Can you try putting the options on the top lvel.
```
loadOptions: {
    method: 'GET',
    headers: {
      'Authorization': 'token ' + Secret("secret-token"),
      'Accept': 'application/vnd.github.v3.raw',
    }
}
```

**Reply by cornhundred** [reply]
Hi @ibgreen I moved it to the top level but I'm seeing the same result from deck.gl

<img width="1103" alt="Screen Shot 2021-02-19 at 1 36 54 PM" src="https://user-images.githubusercontent.com/8352840/108546793-9cca0900-72b7-11eb-93b3-836dd1db9d56.png">

However, the GET request does go through when I use Axios:

<img width="938" alt="Screen Shot 2021-02-19 at 1 39 52 PM" src="https://user-images.githubusercontent.com/8352840/108547048-faf6ec00-72b7-11eb-92cb-36ac92894748.png">

and the network headers from axios' GET request look like 
<img width="939" alt="Screen Shot 2021-02-19 at 1 41 21 PM" src="https://user-images.githubusercontent.com/8352840/108547479-85d7e680-72b8-11eb-972c-514a95eb3045.png">

**Reply by ibgreen** [reply]
OK let's check that you are on latest deck.gl and loaders.gl versions.

Also you probably want to regenerate your access token at this point.

**Reply by cornhundred** [reply]
Token deleted for now :) The versions that I'm using are 

```
deck = require.alias({
  // optional dependencies
  h3: {},
  s2Geometry: {}
})('deck.gl@8.3.7/dist.min.js')
```

and 

```
loaders = require('@loaders.gl/3d-tiles@2.2.8/dist/dist.js')
```

**Reply by kylebarron** [reply]
The latest versions are deck `8.4.6` and loaders `v2.3.12`. You should try with those.

**Reply by cornhundred** [reply]
I was able to upload the deck.gl version, but I'm not sure how to require the latest version of loaders.gl on ObservableHQ (notebook [link](https://observablehq.com/@cornhundred/requiring-deck-gl-and-loaders-gl))

<img width="922" alt="Screen Shot 2021-02-19 at 8 22 54 PM" src="https://user-images.githubusercontent.com/8352840/108578365-5c3ab180-72f0-11eb-89f1-b8411d0472c9.png">

**Reply by cornhundred** [reply]
I'm asking the Observable discourse about how to load the latest version of loaders.gl [here](https://talk.observablehq.com/t/help-requiring-the-latest-version-of-loaders-gl/4679).

**Reply by Pessimistress** [reply]
You do not have to import loaders.gl yourself. The loaders version bundled with the latest deck.gl release was 2.3.11.

Here is a simple proof based on the `TileLayer` documentation example: https://codepen.io/Pessimistress/pen/WNoEjOB

This prop is added to the `TileLayer`:

```js
loadOptions: {
    fetch: {
      headers: {
        'Accept': 'image/png',
        'x-custom-header': 'test'
      }
    }
  },
```

![Screen Shot 2021-02-19 at 6 18 09 PM](https://user-images.githubusercontent.com/2059298/108580158-26410180-72df-11eb-8c68-7e2247ae984e.png)


The problem with your code is that you override the default `getTileData` callback with your own, in which you are not using any fetch options. To fix your `getTileData`, you need to pass the options to your own `load` call:

```js
getTileData: ({x, y, z}) => {
  return loaders.load(`${ROOT_URL}/moon.image_files/${15 + z}/${x}_${y}.jpeg`, {
    fetch: {
      headers: {...}
    }
  });
},
```

**Reply by cornhundred** [reply]
Thanks @Pessimistress for your explanation and help!!! It is working now :D

**Reply by cornhundred** [reply]
@Pessimistress, can I access the loaders that is bundled with deck.gl from the deck.gl object, like I do with the  ScatterplotLayer: `deck.ScatterplotLayer`?

> The loaders version bundled with the latest deck.gl release

**Reply by Pessimistress** [reply]
The deck.gl bundle adds a global object `loaders`.

**Reply by cornhundred** [reply]
Ok, I'm unsure how to access this on ObservableHQ. [This notebook](https://observablehq.com/@cornhundred/deck-gl-moon-tilelayer-example-loaders-global), which used to require loaders at the very bottom cell, but now tries to use the globally defined loaders from deck.gl is not working. [This notebook](https://observablehq.com/@cornhundred/deck-gl-moon-tillayer-example) with the loaders requirement at the bottom is working.

**Reply by Pessimistress** [reply]
If you import deck like this:

```
deck = require.alias({
  // optional dependencies
  h3: {}
})('deck.gl@latest/dist.min.js')
```

In another cell you can do

```
loaders = deck && window.loaders
```

**Reply by cornhundred** [reply]
Thanks @Pessimistress that works :D
